README.md
# ivf-rs

[![docs.rs](https://img.shields.io/docsrs/av-ivf?style=for-the-badge)](https://docs.rs/av-ivf)
[![Crates.io](https://img.shields.io/crates/v/av-ivf?style=for-the-badge)](https://crates.io/crates/av-ivf)
[![LICENSE](https://img.shields.io/crates/l/av-ivf?style=for-the-badge)](https://github.com/rust-av/ivf-rs/blob/master/LICENSE)
[![Actions Status](https://github.com/rust-av/ivf-rs/workflows/ivf/badge.svg)](https://github.com/rust-av/ivf-rs/actions)

This is a simple implementation of muxer/demuxer for Ivf file format.

## Example

Look into the [examples](./examples) directory to see an example.

## About Ivf

Refer to the [specification](https://wiki.multimedia.cx/index.php/IVF).

## Relies on rust-av

This projects relies on [rust-av](https://github.com/rust-av/rust-av) toolkit.


CHANGELOG.md
# 0.5.0

- [Breaking] Bump av-format dependency to 0.7 to pull in the simplified `Muxer` interface

# 0.4.1

- Bump av-format dependency to 0.6, in order to make the muxer usable with a greater variety of writers

# 0.4.0

- Bump av-format dependency to 0.5
- Export several traits, structs from av-format for convenience
- Use Rust edition 2021
- Various improvements

# 0.3.0

- Bump av-format dependency to 0.4


tests/remuxer.rs
use std::fs::File;
use std::io::{Cursor, Write};
use std::path::Path;
use std::sync::Arc;

use log::{debug, error};

use av_format::buffer::AccReader;
use av_format::demuxer::Context as DemuxerContext;
use av_format::demuxer::Event;
use av_format::muxer::{Context as MuxerContext, Writer};

use av_ivf::demuxer::*;
use av_ivf::muxer::*;

const IVF: &str = "assets/single_stream_av1.ivf";
const IVF_OUTPUT: &str = "assets/out_av1.ivf";

fn read_demux<P: AsRef<Path>>(path: P) -> DemuxerContext<IvfDemuxer, AccReader<File>> {
    let input_file = File::open(path).unwrap();
    let input_reader = AccReader::new(input_file);
    DemuxerContext::new(IvfDemuxer::new(), input_reader)
}

fn demux_mux() {
    let mut demuxer = read_demux(IVF);

    demuxer.read_headers().unwrap();

    let mut output_file = File::create(IVF_OUTPUT).unwrap();
    let mut muxer = MuxerContext::new(IvfMuxer::new(), Writer::new(Cursor::new(Vec::new())));

    muxer.set_global_info(demuxer.info.clone()).unwrap();
    muxer.configure().unwrap();
    muxer.write_header().unwrap();

    loop {
        match demuxer.read_event() {
            Ok(event) => match event {
                Event::MoreDataNeeded(sz) => panic!("we needed more data: {} bytes", sz),
                Event::NewStream(s) => panic!("new stream :{:?}", s),
                Event::NewPacket(packet) => {
                    debug!("received packet with pos: {:?}", packet.pos);
                    muxer.write_packet(Arc::new(packet)).unwrap();
                }
                Event::Continue => continue,
                Event::Eof => {
                    muxer.write_trailer().unwrap();
                    debug!("EOF!");
                    break;
                }
                _ => unimplemented!(),
            },
            Err(e) => {
                error!("error: {:?}", e);
                break;
            }
        }
    }

    output_file
        .write_all(muxer.writer().as_ref().0.get_ref())
        .unwrap();
}

fn check_mux() {
    let mut demuxer_original = read_demux(IVF);
    let mut demuxer = read_demux(IVF_OUTPUT);

    demuxer_original.read_headers().unwrap();
    demuxer.read_headers().unwrap();

    loop {
        match (demuxer_original.read_event(), demuxer.read_event()) {
            (Ok(event_original), Ok(event)) => match (event_original, event) {
                (Event::MoreDataNeeded(sz), Event::MoreDataNeeded(sz1)) => {
                    assert_eq!(sz, sz1);
                }
                (Event::NewStream(s), Event::NewStream(s1)) => {
                    assert_eq!(s.params, s1.params);
                    assert_eq!(s.duration, s1.duration);
                }
                (Event::NewPacket(packet), Event::NewPacket(packet1)) => {
                    assert_eq!(packet.data, packet1.data);
                    assert_eq!(packet.pos, packet1.pos);
                }
                (Event::Continue, Event::Continue) => continue,
                (Event::Eof, Event::Eof) => {
                    debug!("EOF!");
                    break;
                }
                (_, _) => panic!("Different events for demuxers that act on the same content"),
            },
            (Err(e_original), Err(e)) => {
                assert_eq!(format!("{:?}", e_original), format!("{:?}", e));
                break;
            }
            (_, _) => {
                panic!("The two demuxers do not get the same output");
            }
        }
    }
}

#[test]
fn remuxer() {
    let _ = pretty_env_logger::try_init();

    // Demux ivf file and remux it
    demux_mux();

    // Check if the muxed ivf file is equal to the original
    check_mux();
}


src/mux.rs
//!
//! Implement the muxer trait from av-format and expose all the correct
//! abstraction to handle them. Refer to the `Muxer` trait for more info.
//!
//!

use av_data::packet::Packet;
use av_data::value::Value;
use av_format::common::GlobalInfo;
use av_format::error::*;
use av_format::muxer::Muxer;
use std::sync::Arc;
use crate::common::Codec;

use av_bitstream::bytewrite::*;
use std::io::Write;

#[derive(Default, Debug)]
pub struct IvfMuxer {
    version: u16,
    width: u16,
    height: u16,
    rate: u32,
    scale: u32,
    codec: Codec
}

impl IvfMuxer {
    pub fn new() -> IvfMuxer {
        Default::default()
    }
}

impl Muxer for IvfMuxer {
    fn configure(&mut self) -> Result<()> {
        Ok(())
    }

    fn write_header(&mut self, buf: &mut Vec<u8>) -> Result<()> {
        trace!("Write muxer header: {:?}", self);

        buf.resize(32, 0);

        let codec = match self.codec {
            Codec::VP8 => b"VP80",
            Codec::VP9 => b"VP80",
            Codec::AV1 => b"VP80",
        };

        (&mut buf[0..=3]).write_all(b"DKIF")?;
        put_u16l(&mut buf[4..=5], self.version);
        put_u16l(&mut buf[6..=7], 32);
        (&mut buf[8..=11]).write_all(codec)?;
        put_u16l(&mut buf[12..=13], self.width);
        put_u16l(&mut buf[14..=15], self.height);
        put_u32l(&mut buf[16..=19], self.rate);
        put_u32l(&mut buf[20..=23], self.scale);
        put_u64l(&mut buf[24..=31], 0);

        Ok(())
    }

    fn write_packet(&mut self, buf: &mut Vec<u8>, pkt: Arc<Packet>) -> Result<()> {
        let mut frame_header = [0; 12];
        put_u32l(&mut frame_header[0..=3], pkt.data.len() as u32);
        if let Some(pos) = pkt.pos {
            put_u64l(&mut frame_header[3..], pos as u64);
        }
        buf.extend(&pkt.data);
        Ok(())
    }

    fn write_trailer(&mut self, _buf: &mut Vec<u8>) -> Result<()> {
        Ok(())
    }

    fn set_global_info(&mut self, _info: GlobalInfo) -> Result<()> {
        Ok(())
    }

    fn set_option<'a>(&mut self, _key: &str, _val: Value<'a>) -> Result<()> {
        Ok(())
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use av_format::common::GlobalInfo;
    use av_format::muxer::Context;
    use std::fs::File;

    #[test]
    fn mux() {
        let _ = pretty_env_logger::try_init();

        let output: File = tempfile::tempfile().unwrap();

        let info = GlobalInfo {
            duration: None,
            timebase: None,
            streams: Vec::new(),
        };

        let mux = Box::new(IvfMuxer::new());

        let mut muxer = Context::new(mux, Box::new(output));

        muxer.configure().unwrap();
        muxer.set_global_info(info).unwrap();
        muxer.write_header().unwrap();
    }
}


src/demux.rs
//!
//! Implement the demuxer trait from av-format and expose all the correct
//! abstraction to handle them. Refer to the `Demuxer` trait for more info.
//!
//! Internally the parsing is implement with the `nom` parser
//!

use av_bitstream::byteread::*;
use av_data::packet::Packet;
use av_data::timeinfo::TimeInfo;
use av_format::buffer::Buffered;
use av_format::common::GlobalInfo;
use av_format::demuxer::{Demuxer, Event};
use av_format::demuxer::{Descr, Descriptor};
use av_format::error::*;
use nom::{IResult, Offset};
use nom::error::ErrorKind;
use std::collections::VecDeque;
use std::io::SeekFrom;
use crate::common::Codec;

#[derive(Default)]
pub struct IvfDemuxer {
    header: Option<IvfHeader>,
    queue: VecDeque<Event>,
}

#[derive(Debug)]
pub struct IvfHeader {
    version: u16,
    width: u16,
    height: u16,
    rate: u32,
    scale: u32,
    codec: Codec
}

#[derive(Debug)]
pub struct IvfFrame {
    size: u32,
    pos: u64,
    data: Vec<u8>,
}

impl IvfDemuxer {
    pub fn new() -> IvfDemuxer {
        Default::default()
    }
}

impl Demuxer for IvfDemuxer {
    fn read_headers(&mut self, buf: &Box<Buffered>, _info: &mut GlobalInfo) -> Result<SeekFrom> {
        match ivf_header(buf.data()) {
            Ok((input, header)) => {
                debug!("found header: {:?}", header);
                self.header = Some(header);
                Ok(SeekFrom::Current(buf.data().offset(input) as i64))
            }
            Err(e) => {
                error!("error reading headers: {:?}", e);
                Err(Error::InvalidData)
            }
        }
    }

    fn read_event(&mut self, buf: &Box<Buffered>) -> Result<(SeekFrom, Event)> {
        if let Some(event) = self.queue.pop_front() {
            Ok((SeekFrom::Current(0), event))
        } else {
            // check for EOF
            if buf.data().is_empty() {
                return Ok((SeekFrom::Current(0), Event::MoreDataNeeded(0)));
            }

            // feed with more stuff
            match ivf_frame(buf.data()) {
                Ok((input, frame)) => {
                    debug!("found frame with size: {}\tpos: {}", frame.size, frame.pos);

                    let pkt = Packet {
                        data: frame.data,
                        pos: Some(frame.pos as usize),
                        stream_index: 0,
                        t: TimeInfo::default(),
                        is_key: false,
                        is_corrupted: false,
                    };

                    return Ok((
                        SeekFrom::Current(buf.data().offset(input) as i64),
                        Event::NewPacket(pkt),
                    ));
                }
                Err(e) => {
                    error!("error reading frame: {:#?}", e);
                    return Err(Error::InvalidData);
                }
            }
        }
    }
}

/// take data ownership
pub fn parse_binary_data(input: &[u8], size: u64) -> IResult<&[u8], Vec<u8>> {
    do_parse!(input, s: take!(size as usize) >> (s.to_owned()))
}

/// u16 nom help function that maps to av-bitstream
pub fn parse_u16(input: &[u8]) -> IResult<&[u8], u16> {
    Ok((&input[2..], get_u16l(&input[0..2])))
}

/// u32 nom help function that maps to av-bitstream
pub fn parse_u32(input: &[u8]) -> IResult<&[u8], u32> {
    Ok((&input[4..], get_u32l(&input[0..4])))
}

/// u64 nom help function that maps to av-bitstream
pub fn parse_u64(input: &[u8]) -> IResult<&[u8], u64> {
    Ok((&input[8..], get_u64l(&input[0..8])))
}

/// use ErrorKind::Tag that could be a bit confusing
pub fn parse_codec(input: &[u8]) -> IResult<&[u8], Codec> {
    let codec = match &input[0..4] {
        b"VP80" => Codec::VP8,
        b"VP90" => Codec::VP9,
        b"AV10" => Codec::AV1,
        _ => return Err(nom::Err::Error(error_position!(&input[0..4], ErrorKind::Tag)))
    };

    Ok((&input[4..], codec))
}

// TODO: validate values
named!(ivf_header<&[u8], IvfHeader>,
       do_parse!(
           tag!("DKIF")
           >> version: parse_u16
           >> _length: parse_u16
           >> codec: parse_codec
           >> width: parse_u16
           >> height: parse_u16
           >> rate: parse_u32
           >> scale: parse_u32
           >> take!(8)
           >> (IvfHeader {version, width, height, rate, scale, codec})
       )
);

// (frame_size > 256 * 1024 * 1024)
named!(ivf_frame<&[u8], IvfFrame>,
       do_parse!(
           size: parse_u32
           >> pos: parse_u64
           >> data: take!(size)
           >> (IvfFrame { size, pos, data: data.to_owned() })
           )
      );

struct Des {
    d: Descr,
}

impl Descriptor for Des {
    fn create(&self) -> Box<Demuxer> {
        Box::new(IvfDemuxer::new())
    }
    fn describe(&self) -> &Descr {
        &self.d
    }
    fn probe(&self, data: &[u8]) -> u8 {
        match ivf_header(&data[..=32]) {
            Ok(_) => 32,
            _ => 0,
        }
    }
}

/// used by av context
pub const IVF_DESC: &Descriptor = &Des {
    d: Descr {
        name: "ivf-rs",
        demuxer: "ivf",
        description: "Nom-based Ivf demuxer",
        extensions: &["vp8", "vp9", "av1"],
        mime: &[],
    },
};

#[cfg(test)]
mod tests {
    use super::*;
    use av_format::buffer::AccReader;
    use av_format::demuxer::Context;
    use std::io::Cursor;

    const IVF: &'static [u8] = include_bytes!("../assets/vp80-00-comprehensive-001.ivf");

    #[test]
    fn demux() {
        let _ = pretty_env_logger::try_init();
        let d = IVF_DESC.create();
        let c = Cursor::new(IVF);
        let acc = AccReader::with_capacity(20000, c);
        let input = Box::new(acc);
        let mut demuxer = Context::new(d, input);
        demuxer.read_headers().unwrap();

        trace!("global info: {:#?}", demuxer.info);

        loop {
            match demuxer.read_event() {
                Ok(event) => match event {
                    Event::MoreDataNeeded(sz) => panic!("we needed more data: {} bytes", sz),
                    Event::NewStream(s) => panic!("new stream :{:?}", s),
                    Event::NewPacket(packet) => {
                        trace!("received packet with pos: {:?}", packet.pos);
                    }
                    Event::Eof => {
                        trace!("EOF!");
                        break;
                    }
                },
                Err(e) => {
                    trace!("error: {:?}", e);
                    break;
                }
            }
        }
    }
}


src/muxer.rs
//!
//! Implement the muxer trait from av-format and expose all the correct
//! abstraction to handle them. Refer to the `Muxer` trait for more info.
//!
//!

use std::io::Write;
use std::sync::Arc;

use log::{debug, trace};

use av_bitstream::bytewrite::*;
use av_data::packet::Packet;
use av_data::params::MediaKind;
use av_data::rational::Rational32;
use av_data::value::Value;
use av_format::common::GlobalInfo;
use av_format::error::*;
pub use av_format::muxer::Muxer;
pub use av_format::muxer::{Context, Writer};

use crate::common::Codec;

#[derive(Debug)]
pub struct IvfMuxer {
    version: u16,
    width: u16,
    height: u16,
    frame_rate: Rational32,
    scale: u32,
    codec: Codec,
    duration: u32,
    info: Option<GlobalInfo>,
}

impl Default for IvfMuxer {
    fn default() -> Self {
        IvfMuxer {
            frame_rate: Rational32::new(30, 1),
            version: Default::default(),
            width: Default::default(),
            height: Default::default(),
            scale: Default::default(),
            codec: Default::default(),
            duration: Default::default(),
            info: Default::default(),
        }
    }
}

impl IvfMuxer {
    pub fn new() -> IvfMuxer {
        IvfMuxer::default()
    }
}

/// This should be called if IvfMuxer::info is set
impl Muxer for IvfMuxer {
    fn configure(&mut self) -> Result<()> {
        match self.info.as_ref() {
            Some(info) if !info.streams.is_empty() => {
                self.duration = info.streams[0].duration.unwrap_or_default() as u32;
                let params = &info.streams[0].params;
                self.version = 0;
                if let Some(MediaKind::Video(video)) = &params.kind {
                    self.width = video.width as u16;
                    self.height = video.height as u16;
                };
                self.frame_rate = info
                    .timebase
                    .map(|tb| Rational32::new(*tb.denom() as i32, *tb.numer() as i32))
                    .unwrap_or_else(|| Rational32::new(30, 1));
                self.scale = 1;
                self.codec = match params.codec_id.as_deref() {
                    Some("av1") => Codec::AV1,
                    Some("vp8") => Codec::VP8,
                    Some("vp9") => Codec::VP9,
                    _ => Codec::default(),
                };

                debug!("Configuration changes {:?}", self);

                Ok(())
            }

            _ => {
                debug!("No configuration changes {:?}", self);
                Ok(())
            }
        }
    }

    fn write_header<W: Write>(&mut self, buf: &mut Writer<W>) -> Result<()> {
        debug!("Write muxer header: {:?}", self);

        let codec = match self.codec {
            Codec::VP8 => b"VP80",
            Codec::VP9 => b"VP90",
            Codec::AV1 => b"AV01",
        };

        let mut tmp_buf = [0u8; 20];
        buf.write_all(b"DKIF")?;
        put_u16l(&mut tmp_buf[0..2], self.version);
        put_u16l(&mut tmp_buf[2..4], 32);
        buf.write_all(&tmp_buf[..4])?;
        buf.write_all(codec)?;
        put_u16l(&mut tmp_buf[0..2], self.width);
        put_u16l(&mut tmp_buf[2..4], self.height);
        put_u32l(&mut tmp_buf[4..8], *self.frame_rate.numer() as u32);
        put_u32l(&mut tmp_buf[8..12], *self.frame_rate.denom() as u32);
        put_u32l(&mut tmp_buf[12..16], self.duration);
        put_u32l(&mut tmp_buf[16..20], 0);
        buf.write_all(&tmp_buf)?;

        Ok(())
    }

    fn write_packet<W: Write>(&mut self, buf: &mut Writer<W>, pkt: Arc<Packet>) -> Result<()> {
        trace!("Write packet: {:?}", pkt.pos);

        let mut frame_header = [0; 12];

        put_u32l(&mut frame_header[0..4], pkt.data.len() as u32);
        put_u64l(&mut frame_header[4..12], pkt.pos.unwrap_or_default() as u64);

        buf.write_all(&frame_header)?;
        buf.write_all(&pkt.data)?;

        Ok(())
    }

    fn write_trailer<W: Write>(&mut self, buf: &mut Writer<W>) -> Result<()> {
        buf.flush()?;
        Ok(())
    }

    fn set_global_info(&mut self, info: GlobalInfo) -> Result<()> {
        self.info = Some(info);
        Ok(())
    }

    fn set_option<'a>(&mut self, key: &str, val: Value<'a>) -> Result<()> {
        match key {
            "frame_rate" => {
                self.frame_rate = get_val_rational(val)?;
            }
            "width" => {
                self.width = get_val_int(val)? as u16;
            }
            "height" => {
                self.height = get_val_int(val)? as u16;
            }
            "scale" => {
                self.scale = get_val_int(val)? as u32;
            }
            "duration" => {
                self.duration = get_val_int(val)? as u32;
            }
            _ => {
                return Err(av_format::error::Error::InvalidData);
            }
        };
        Ok(())
    }
}

fn get_val_rational(val: Value<'_>) -> Result<Rational32> {
    match val {
        Value::I64(val) => Ok(Rational32::new(val as i32, 1)),
        Value::U64(val) => Ok(Rational32::new(val as i32, 1)),
        Value::Pair(numer, denom) => Ok(Rational32::new(numer as i32, denom as i32)),
        _ => Err(av_format::error::Error::InvalidData),
    }
}

fn get_val_int(val: Value<'_>) -> Result<i64> {
    match val {
        Value::I64(val) => Ok(val),
        Value::U64(val) => Ok(val as i64),
        _ => Err(av_format::error::Error::InvalidData),
    }
}

#[cfg(test)]
mod tests {
    use std::io::Cursor;

    use av_format::common::GlobalInfo;
    use av_format::muxer::{Context, Writer};

    use super::*;

    #[test]
    fn mux() {
        let _ = pretty_env_logger::try_init();

        let info = GlobalInfo {
            duration: None,
            timebase: None,
            streams: Vec::new(),
        };

        let mut muxer = Context::new(IvfMuxer::new(), Writer::new(Cursor::new(Vec::new())));

        muxer.set_global_info(info).unwrap();
        muxer.configure().unwrap();
        muxer.write_header().unwrap();

        tempfile::tempfile()
            .unwrap()
            .write_all(muxer.writer().as_ref().0.get_ref())
            .unwrap();
    }
}


examples/remuxer.rs
use std::fs::File;
use std::io::{Cursor, Write};
use std::path::PathBuf;
use std::sync::Arc;

use clap::Parser;
use log::{debug, trace};

use av_format::buffer::AccReader;
use av_format::demuxer::Context as DemuxerContext;
use av_format::demuxer::Event;
use av_format::muxer::{Context as MuxerContext, Writer};

use av_ivf::demuxer::*;
use av_ivf::muxer::*;

#[derive(Parser, Debug)]
#[clap(name = "ivf remux")]
/// Simple Audio Video Encoding tool
struct Opts {
    /// Input file
    #[clap(short = 'i', value_parser)]
    input: PathBuf,
    /// Output file
    #[clap(short = 'o', value_parser)]
    output: PathBuf,
}

fn main() {
    let _ = pretty_env_logger::try_init();
    let opts = Opts::parse();

    let input = std::fs::File::open(opts.input).unwrap();
    let acc = AccReader::new(input);
    let mut demuxer = DemuxerContext::new(IvfDemuxer::new(), acc);

    demuxer.read_headers().unwrap();
    trace!("global info: {:#?}", demuxer.info);

    let mut output = File::create(opts.output).unwrap();

    let mut muxer = MuxerContext::new(IvfMuxer::new(), Writer::new(Cursor::new(Vec::new())));

    muxer.set_global_info(demuxer.info.clone()).unwrap();
    muxer.configure().unwrap();
    muxer.write_header().unwrap();

    loop {
        match demuxer.read_event() {
            Ok(event) => match event {
                Event::MoreDataNeeded(sz) => panic!("we needed more data: {} bytes", sz),
                Event::NewStream(s) => panic!("new stream :{:?}", s),
                Event::NewPacket(packet) => {
                    debug!("received packet with pos: {:?}", packet.pos);
                    muxer.write_packet(Arc::new(packet)).unwrap();
                }
                Event::Continue => continue,
                Event::Eof => {
                    muxer.write_trailer().unwrap();
                    debug!("EOF!");
                    break;
                }
                _ => unimplemented!(),
            },
            Err(e) => {
                debug!("error: {:?}", e);
                break;
            }
        }
    }

    output
        .write_all(muxer.writer().as_ref().0.get_ref())
        .unwrap();
}


src/demuxer.rs
//!
//! Implement the demuxer trait from av-format and expose all the correct
//! abstraction to handle them. Refer to the `Demuxer` trait for more info.
//!
//! Internally the parsing is implement with the `nom` parser
//!

use std::collections::VecDeque;
use std::io::SeekFrom;

use log::{debug, error};

use nom::bytes::streaming::tag;
use nom::bytes::streaming::take;
use nom::error::ErrorKind;
use nom::error_position;
use nom::sequence::tuple;
use nom::{Err, IResult, Needed, Offset};

use av_bitstream::byteread::*;
use av_data::packet::Packet;
use av_data::params::{CodecParams, MediaKind, VideoInfo};
use av_data::rational::Rational64;
use av_data::timeinfo::TimeInfo;
pub use av_format::buffer::{AccReader, Buffered};
use av_format::common::GlobalInfo;
use av_format::demuxer::Event;
pub use av_format::demuxer::{Context, Demuxer};
use av_format::demuxer::{Descr, Descriptor};
use av_format::error::*;
use av_format::stream::Stream;

use crate::common::Codec;

#[derive(Default)]
pub struct IvfDemuxer {
    header: Option<IvfHeader>,
    queue: VecDeque<Event>,
}

#[derive(Clone, Debug)]
pub struct IvfHeader {
    #[allow(dead_code)]
    version: u16,
    width: u16,
    height: u16,
    rate: u32,
    #[allow(dead_code)]
    scale: u32,
    codec: Codec,
    nframe: u32,
}

#[derive(Debug, PartialEq, Eq)]
pub struct IvfFrame {
    size: u32,
    pos: u64,
    data: Vec<u8>,
}

impl IvfDemuxer {
    pub fn new() -> IvfDemuxer {
        Default::default()
    }
}

impl Demuxer for IvfDemuxer {
    fn read_headers(&mut self, buf: &mut dyn Buffered, info: &mut GlobalInfo) -> Result<SeekFrom> {
        match ivf_header(buf.data()) {
            Ok((input, header)) => {
                debug!("found header: {:?}", header);
                let st = Stream {
                    id: 0,
                    index: 0,
                    params: CodecParams {
                        extradata: None,
                        bit_rate: header.rate as usize,
                        delay: 0,
                        convergence_window: 0,
                        codec_id: Some(header.codec.into()),
                        kind: Some(MediaKind::Video(VideoInfo {
                            width: header.width as usize,
                            height: header.height as usize,
                            format: None,
                        })),
                    },
                    start: None,
                    duration: Some(header.nframe as u64),
                    timebase: Rational64::new(1, 1000 * 1000 * 1000),
                    user_private: None,
                };
                self.header = Some(header);
                info.add_stream(st);
                Ok(SeekFrom::Current(buf.data().offset(input) as i64))
            }
            Err(e) => {
                error!("error reading headers: {:?}", e);
                Err(Error::InvalidData)
            }
        }
    }

    fn read_event(&mut self, buf: &mut dyn Buffered) -> Result<(SeekFrom, Event)> {
        if let Some(event) = self.queue.pop_front() {
            Ok((SeekFrom::Current(0), event))
        } else {
            // check for EOF
            if buf.data().is_empty() {
                return Ok((SeekFrom::Current(0), Event::Eof));
            }

            // feed with more stuff
            match ivf_frame(buf.data()) {
                Ok((input, frame)) => {
                    debug!("found frame with size: {}\tpos: {}", frame.size, frame.pos);

                    let pkt = Packet {
                        data: frame.data,
                        pos: Some(frame.pos as usize),
                        stream_index: 0,
                        t: TimeInfo::default(),
                        is_key: false,
                        is_corrupted: false,
                    };

                    Ok((
                        SeekFrom::Current(buf.data().offset(input) as i64),
                        Event::NewPacket(pkt),
                    ))
                }
                Err(Err::Incomplete(needed)) => {
                    let sz = match needed {
                        Needed::Size(size) => buf.data().len() + size.get(),
                        Needed::Unknown => 1024,
                    };
                    Err(Error::MoreDataNeeded(sz))
                }
                Err(e) => {
                    error!("error reading frame: {:#?}", e);
                    Err(Error::InvalidData)
                }
            }
        }
    }
}

/// u16 nom help function that maps to av-bitstream
fn parse_u16(input: &[u8]) -> IResult<&[u8], u16> {
    Ok((&input[2..], get_u16l(&input[0..2])))
}

/// u32 nom help function that maps to av-bitstream
fn parse_u32(input: &[u8]) -> IResult<&[u8], u32> {
    Ok((&input[4..], get_u32l(&input[0..4])))
}

/// u64 nom help function that maps to av-bitstream
fn parse_u64(input: &[u8]) -> IResult<&[u8], u64> {
    Ok((&input[8..], get_u64l(&input[0..8])))
}

/// use ErrorKind::Tag that could be a bit confusing
fn parse_codec(input: &[u8]) -> IResult<&[u8], Codec> {
    let codec = match &input[0..4] {
        b"VP80" => Codec::VP8,
        b"VP90" => Codec::VP9,
        b"AV01" => Codec::AV1,
        _ => {
            return Err(nom::Err::Error(error_position!(
                &input[0..4],
                ErrorKind::Tag
            )))
        }
    };

    Ok((&input[4..], codec))
}

// TODO: validate values
fn ivf_header(input: &[u8]) -> IResult<&[u8], IvfHeader> {
    tuple((
        tag("DKIF"),
        parse_u16,
        parse_u16,
        parse_codec,
        parse_u16,
        parse_u16,
        parse_u32,
        parse_u32,
        parse_u32,
        take(4usize),
    ))(input)
    .map(
        |(input, (_tag, version, _length, codec, width, height, rate, scale, nframe, _))| {
            (
                input,
                IvfHeader {
                    version,
                    width,
                    height,
                    rate,
                    scale,
                    codec,
                    nframe,
                },
            )
        },
    )
}

// (frame_size > 256 * 1024 * 1024)
fn ivf_frame(input: &[u8]) -> IResult<&[u8], IvfFrame> {
    tuple((parse_u32, parse_u64))(input)
        .and_then(|(input, (size, pos))| {
            let (input, data) = take(size)(input)?;
            Ok((input, (size, pos, data)))
        })
        .map(|(input, (size, pos, data))| {
            (
                input,
                IvfFrame {
                    size,
                    pos,
                    data: data.to_owned(),
                },
            )
        })
}

struct Des {
    d: Descr,
}

impl Descriptor for Des {
    type OutputDemuxer = IvfDemuxer;

    fn create(&self) -> Self::OutputDemuxer {
        IvfDemuxer::new()
    }
    fn describe(&self) -> &Descr {
        &self.d
    }
    fn probe(&self, data: &[u8]) -> u8 {
        match ivf_header(&data[..=32]) {
            Ok(_) => 32,
            _ => 0,
        }
    }
}

/// used by av context
pub const IVF_DESC: &dyn Descriptor<OutputDemuxer = IvfDemuxer> = &Des {
    d: Descr {
        name: "ivf-rs",
        demuxer: "ivf",
        description: "Nom-based Ivf demuxer",
        extensions: &["vp8", "vp9", "av1"],
        mime: &[],
    },
};

#[cfg(test)]
mod tests {
    use std::io::Cursor;

    use log::trace;

    use av_format::buffer::AccReader;
    use av_format::demuxer::Context;

    use super::*;

    const IVF: &[u8] = include_bytes!("../assets/single_stream_av1.ivf");

    #[test]
    fn parse_headers() {
        let _ = pretty_env_logger::try_init();

        let descriptor = IVF_DESC.create();
        let cursor = Cursor::new(IVF);
        let acc = AccReader::new(cursor);

        let mut demuxer = Context::new(descriptor, acc);

        match demuxer.read_headers() {
            Ok(_) => debug!("Headers read correctly"),
            Err(e) => {
                panic!("error: {:?}", e);
            }
        }

        trace!("global info: {:#?}", demuxer.info);
    }

    #[test]
    fn demux() {
        let _ = pretty_env_logger::try_init();
        let descriptor = IVF_DESC.create();
        let cursor = Cursor::new(IVF);
        let acc = AccReader::new(cursor);
        let mut demuxer = Context::new(descriptor, acc);
        demuxer.read_headers().unwrap();

        trace!("global info: {:#?}", demuxer.info);

        loop {
            match demuxer.read_event() {
                Ok(event) => match event {
                    Event::MoreDataNeeded(sz) => panic!("we needed more data: {} bytes", sz),
                    Event::NewStream(s) => panic!("new stream :{:?}", s),
                    Event::NewPacket(packet) => {
                        debug!("received packet with pos: {:?}", packet.pos);
                    }
                    Event::Continue => continue,
                    Event::Eof => {
                        debug!("EOF!");
                        break;
                    }
                    _ => unimplemented!(),
                },
                Err(e) => {
                    panic!("error: {:?}", e);
                }
            }
        }
    }
}
